lib/checksum.c

```diff
@@ -42,794 +41,858 @@
+                              paranoid->twin_candidate, paranoid);
 #endif
-                if(!paranoid->shadow_hash) {
-                    /* we use the rejects file to speed up rm_digest_equal */
-                    paranoid->rejects =
-                        g_slist_prepend(paranoid->rejects, paranoid->twin_candidate);
-                }
-                paranoid->twin_candidate = NULL;
-                paranoid->twin_candidate_buffer = NULL;
-            } else {
+            if(!paranoid->shadow_hash) {
+                /* we use the rejects file to speed up rm_digest_equal */
+                paranoid->rejects =
+                    g_slist_prepend(paranoid->rejects, paranoid->twin_candidate);
+            }
+            paranoid->twin_candidate = NULL;
+            paranoid->twin_candidate_buffer = NULL;
+        } else {
 #if _RM_CHECKSUM_DEBUG
-                rm_log_debug_line("Added twin candidate %p for %p",
-                                  paranoid->twin_candidate, paranoid);
+            rm_log_debug_line("Added twin candidate %p for %p", paranoid->twin_candidate,
+                              paranoid);
 #endif
-            }
         }
     }
 }
 
-#define BLAKE_COPY(ALGO)                                                        \
-    {                                                                           \
-        self = g_slice_new0(RmDigest);                                          \
-        self->bytes = digest->bytes;                                            \
-        self->type = digest->type;                                              \
-        self->ALGO##_state = g_slice_alloc0(sizeof(ALGO##_state));              \
-        memcpy(self->ALGO##_state, digest->ALGO##_state, sizeof(ALGO##_state)); \
+static void rm_digest_paranoid_steal(RmParanoid *paranoid, guint8 *result) {
+    RmDigest *shadow_hash = paranoid->shadow_hash;
+    rm_digest_xxhash_steal(shadow_hash->state, result);
+}
+
+/* Note: paranoid update implementation is in rm_digest_buffered_update() below */
+
+static const RmDigestInterface paranoid_interface = {
+    .name = "paranoid",
+    .bits = 64, /* must match shadow hash length */
+    .len = NULL,
+    .new = (RmDigestNewFunc)rm_digest_paranoid_new,
+    .free = (RmDigestFreeFunc)rm_digest_paranoid_free,
+    .update = NULL,
+    .copy = NULL,
+    .steal = (RmDigestStealFunc)rm_digest_paranoid_steal};
+
+////////////////////////////////
+//   RmDigestInterface map    //
+////////////////////////////////
+
+static const RmDigestInterface *rm_digest_get_interface(RmDigestType type) {
+    static const RmDigestInterface *digest_interfaces[] = {
+        [RM_DIGEST_UNKNOWN] = NULL,
+        [RM_DIGEST_MURMUR] = &murmur_interface,
+        [RM_DIGEST_METRO] = &metro_interface,
+        [RM_DIGEST_METRO256] = &metro256_interface,
+#if HAVE_SSE_4_2
+        [RM_DIGEST_METROCRC] = &metrocrc_interface,
+        [RM_DIGEST_METROCRC256] = &metrocrc256_interface,
```

Comment:
Similar to the 32 bit comment: Did you try to compile without sse4.2? We had a nasty bug, we're the official arch package was built with sse4.2 support and thus it crashed violently for some (me) that didn't have the instruction set.

Reply:
That's an interesting dilemma for package-manager-managed releases.  It seems there is a runtime check (`__builtin_cpu_supports ("sse4.2"`) so maybe I'll rewrite that way.